core/secrets/secret.go

```diff
@@ -0,0 +1,88 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+package secrets
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/juju/errors"
+)
+
+// Scope represents the scope of a secret.
+type Scope string
+
+const (
+	ScopeApplication = Scope("application")
+)
+
+var validScopes = map[Scope]bool{
+	ScopeApplication: true,
+}
+
+// SecretType is the type of a secret.
+// This is used when creating a secret.
+type SecretType string
+
+const (
+	TypeBlob     = SecretType("blob")
+	TypePassword = SecretType("password")
+)
+
+var validTypes = map[SecretType]bool{
+	TypeBlob:     true,
+	TypePassword: true,
+}
+
+// SecretConfig is used when cresting a secret.
+type SecretConfig struct {
+	Type   SecretType
+	Path   string
+	Scope  Scope
+	Params map[string]interface{}
+}
+
+// NewSecretConfig is used to create an application scoped blob secret.
+func NewSecretConfig(nameParts ...string) *SecretConfig {
+	return &SecretConfig{
+		Type:  TypeBlob,
+		Scope: ScopeApplication,
+		Path:  strings.Join(nameParts, "."),
+	}
+}
+
+// TODO(wallyworld) - use a schema to describe the config
+const (
+	PasswordLength       = "password-length"
+	PasswordSpecialChars = "password-special-chars"
+)
+
+// NewPasswordSecretConfig is used to create an application scoped password secret.
+func NewPasswordSecretConfig(length int, specialChars bool, nameParts ...string) *SecretConfig {
+	return &SecretConfig{
+		Type:  TypePassword,
+		Scope: ScopeApplication,
+		Path:  strings.Join(nameParts, "."),
+		Params: map[string]interface{}{
+			PasswordLength:       length,
+			PasswordSpecialChars: specialChars,
+		},
+	}
+}
+
+var pathRegexp = regexp.MustCompile("^[a-zA-Z0-9]+(\\.[a-zA-Z0-9]+)*$")
```

Comment:
Might be better specified with a backtick-string to avoid the double `\\` (regex literals normally are).